hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationMarkerChore.java

```diff
@@ -0,0 +1,118 @@
+ * {@link #REPLICATION_MARKER_ENABLED_KEY} is set to true, then we will create 1 marker row every
+ * {@link #REPLICATION_MARKER_CHORE_DURATION_KEY} ms
+ * {@link org.apache.hadoop.hbase.replication.regionserver.ReplicationSourceWALReader} will populate
+ * the Replication Marker edit with region_server_name, wal_name and wal_offset encoded in
+ * {@link org.apache.hadoop.hbase.shaded.protobuf.generated.WALProtos.ReplicationMarkerDescriptor}
+ * object. {@link org.apache.hadoop.hbase.replication.regionserver.Replication} will change the
+ * REPLICATION_SCOPE for this edit to GLOBAL so that it can replicate. On the sink cluster,
+ * {@link org.apache.hadoop.hbase.replication.regionserver.ReplicationSink} will convert the
+ * ReplicationMarkerDescriptor into a Put mutation to REPLICATION_SINK_TRACKER_TABLE_NAME_STR table.
+ */
+@InterfaceAudience.Private
+public class ReplicationMarkerChore extends ScheduledChore {
+  private static final Logger LOG = LoggerFactory.getLogger(ReplicationMarkerChore.class);
+  private static final MultiVersionConcurrencyControl MVCC = new MultiVersionConcurrencyControl();
+  public static final RegionInfo REGION_INFO =
+    RegionInfoBuilder.newBuilder(REPLICATION_SINK_TRACKER_TABLE_NAME).build();
+  private static final String DELIMITER = "_";
+  private final Configuration conf;
+  private final RegionServerServices rsServices;
+  private WAL wal;
+
+  public static final String REPLICATION_MARKER_ENABLED_KEY =
+    "hbase.regionserver.replication.marker.enabled";
+  public static final boolean REPLICATION_MARKER_ENABLED_DEFAULT = false;
+
+  public static final String REPLICATION_MARKER_CHORE_DURATION_KEY =
+    "hbase.regionserver.replication.marker.chore.duration";
+  public static final int REPLICATION_MARKER_CHORE_DURATION_DEFAULT = 30 * 1000; // 30 seconds
+
+  public ReplicationMarkerChore(final Stoppable stopper, final RegionServerServices rsServices,
+    int period, Configuration conf) {
+    super("ReplicationTrackerChore", stopper, period);
+    this.conf = conf;
+    this.rsServices = rsServices;
+  }
+
+  @Override
+  protected void chore() {
+    if (wal == null) {
+      try {
+        wal = rsServices.getWAL(null);
+      } catch (IOException ioe) {
+        LOG.warn("Unable to get WAL ", ioe);
+        // Shouldn't happen. Ignore and wait for the next chore run.
+        return;
+      }
+    }
+    String serverName = rsServices.getServerName().getServerName();
+    long timeStamp = EnvironmentEdgeManager.currentTime();
+    // We only have timestamp in ReplicationMarkerDescriptor and the remaining properties walname,
+    // regionserver name and wal offset at ReplicationSourceWALReaderThread.
+    byte[] rowKey = getRowKey(serverName, timeStamp);
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Creating replication marker edit.");
+    }
+    try {
+      WALUtil.writeReplicationMarkerAndSync(wal, MVCC, REGION_INFO, rowKey, timeStamp);
```

Comment:
> Ned to make sure that the system table is never split. Checked the creator code above, seems we do not set a special split policy for the system table?

Didn't understand this comment. We don't persist this marker row in the source cluster. We replicate this marker row and persist it in the sink cluster. On the sink side (ReplicationSink.java), we just extract the table name from this REGION_INFO object.

> If we enable multi wal, then here we can only add markers to one WAL group, the WAL files in other group will not have markers.

I haven't tested this feature with multi wal implementation. Can we work on enabling this feature for multi wal once we merge this PR ?

Reply:
If we only use table name in this method, then better just pass a table name in. The use of a hard coded REGION_INFO implies that there is only one region for this table. In HBase, we hard coded the meta region info, RegionInfoBuilder.FIRST_META_REGIONINFO. So if the table is splittable, then here let's not use hard coded region info.

Reply:
And for the multi wal support, we do not need to implement it in this PR. But let's at least discuss about the design and make sure that we do not need to do breaking changes when we want to add the support later. And also, if the feature will be released before we add the multi wal support, we need to find a way to make sure that user can not enable multi wal and this feature together.

Reply:
> And for the multi wal support, we do not need to implement it in this PR. But let's at least discuss about the design and make sure that we do not need to do breaking changes when we want to add the support later. 

What I was thinking is as follows: 
1. Currently `ReplicationMarkerChore` calls  `rsServices.getWAL` to get the current WAL. In multi WAL implementation, we will replace it with `rsServices.getWALs` and iterate over all the WALs and create a replication marker row for each WAL. This will seamlessly handled via ReplicationSourceWALReader class to replicate.
@Apache9 WDYT ?

> And also, if the feature will be released before we add the multi wal support, we need to find a way to make sure that user can not enable multi wal and this feature together.

I think it is fine keeping this feature with multi wal support also. I tested the test case `TestReplicationMarker` with provider as `multiwal` and it worked fine. However it will create marker row for just first WAL but since we use this feature just for persisting events to a table, it doesn't cause any harm.
@Apache9  WDYT ?

Reply:
> > If we only use table name in this method, then better just pass a table name in.
> 
> All the methods which are used to sync the marker rows expect regioninfo parameter. For e.g `WALUtil#writeMarker` [here](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/WALUtil.java#L139) Also the API to append marker row to WAL uses regioninfo. Refer [here](https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/wal/AbstractFSWAL.java#L1257)

That's the problem here... We need to pass a region info in, so we hard coded a region info, which implies that this table only has one region, while it is actually not...

And this makes me a bit worry that, actually we do not hold this region on the region server, but then we write a WAL which is for this region, I'm afraid in the future if other developers add some sanity checks in the WAL implementation and then it will break this feature...

Reply:
Skimmed the code, our WAL is only designed to write WAL entries which are bound to a region, but here, actually we want to write a region free marker, right? So I think we'd better have a more high level design on how this is expected to work, not only 'OK, it works'. For example, the region info will be passed to coprocessor, it may confuse the develop who implements the coprocessor as the region is not on the region server, and even does not exist...

Reply:
> For example, the region info will be passed to coprocessor, it may confuse the develop who implements the coprocessor as the region is not on the region server, and even does not exist...

The following co-processor hooks are invoked from the time we append the entry to WAL until we create a Put request on the destination side. Please let me know if I missed anything.
```
WALCoprocessorHost#preWALWrite
WALCoprocessorHost#postWALWrite
```
Can we just add a condition to return early if the edit belongs to METAFAMILY and REPLICATION_MARKER qualifier ? 
@Apache9  WDYT ? or let me know if you have a different idea.

Reply:
@Apache9  Any comments ?

Reply:
The main problem here is whether we allow writing WAL entries which are not bound to a region. In the current region server architecture, when we want to get a WAL instance, we need to have a region right?

So this is the main concern here. Either we change the abstraction to allow writing WAL entries which are not bound to a region, or we change the the implementation here, for example, maybe attach the marker to an existing WAL entry?

Reply:
> Either we change the abstraction to allow writing WAL entries which are not bound to a region, or we change the the implementation here, for example, maybe attach the marker to an existing WAL entry?

Exactly.

Reply:
> Either we change the abstraction to allow writing WAL entries which are not bound to a region,

As mentioned in my previous comment [here](https://github.com/apache/hbase/pull/4556/files#r911241234), I think we can just override `WALCoprocessorHost#preWALWrite` and `WALCoprocessorHost#postWALWrite` to skip running coproc hooks if the edit is of type "REPLICATION_MARKER". @Apache9  WDYT ?

Reply:
Skipping does not solve the actual problem, we should have an overall design, on how to allow writing WAL entries which are not bound to a region(or we still do not allow these type of WAL entries and make our markers part of a normal WAL entry). This is the key problem here. The current implementation is a bit hack, which may have long term impact.

Reply:
@Apache9 thanks for your inputs on this thread. Could you provide an example of what you think would not be considered a hack for the current design? Do you think everything about the marker (not bound to a region) needs to change or something specific here?

Reply:
I have created a new class called `NoRegionWALEdit` to create an edit which is not bound to any region. Within `WALCoprocessorHost#preWrite` and `WALCoprocessorHost#postWrite` methods we can check if the edit is an instance of `NoRegionWALEdit`. If yes then we can bypass these coproc hooks and we can add a this check in custom co-procs also. See commit [here](https://github.com/apache/hbase/pull/4556/commits/7126cfb7ddd298ee66fb21e02d59d8d5d3688670).
@Apache9  Would this be sufficient to move this PR forward. If not, then please provide an _actionable_ feedback. I would like to make progress on this PR. Thank you !

Reply:
I suggest we go back to the start of the tech design part. Here the problem is we write a WAL edit which belongs to a region which is not on the region server, we can hardly say this is a valid operation.

For me, I think there are two choices.
First, we could try to make HBase accept WAL edits which do not belong to any regions. For example, we introduce a new WAL type called 'ServerLevelMarker', give an overall design on how we plan to deal with these WAL edits in HBase, and then we use this 'ServerLevelMarker' to implement the feature here.

Second, we could try to attach our markers to valid WAL edits, resue the METAFAMILY mechanism to filter them out and not let them mess up the normal WAL processing.

Thanks.

Reply:
Thank you you @Apache9  for your reply. I was busy with some other high priority work at my day job so couldn't find time to reply here or update the patch.

> Second, we could try to attach our markers to valid WAL edits, resue the METAFAMILY mechanism to filter them out and not let them mess up the normal WAL processing.

I like this suggestion where ReplicationMarkerChore will choose one region at random among the regions it hosts. It creates a replication marker edit with that region. We reuse the METAFAMILY mechanism to filter them out at various places (like ReplicationSourceWALActionsListener and replication related code) and replicate it to the sink cluster. Added a new commit here https://github.com/apache/hbase/pull/4556/commits/d4977208acfdff79c778ec23fd4cc276b53a176c
Can you please review if this approach is clean ? Thank you.